### PR TITLE
add support for the HLG EOTF and a display model for the iPhone 14 Pro with HDR to test it

### PR DESCRIPTION
**Pull request by cosmin**
~~This borrows the EOTF from VQEG `siti-tools` at https://github.com/VQEG/siti-tools/blob/main/siti_tools/siti.py#L471 and adapts it to PyTorch. I haven't actually checked the math but presumably the formula in siti-tools is correct.~~

Update: the application of the EOTF from siti-tools was producing some questionable scores on dark scenes. I went ahead and derived the EOTF directly from Table 5 of https://www.itu.int/dms_pubrec/itu-r/rec/bt/R-REC-BT.2100-1-201706-S!!PDF-E.pdf and most importantly replaced the application of `gamma - 1` with `gamma` directly (and updated the calculation to use the more precise formula from 5e to account for the > 1,000 nit iPhone display).

The application of gamma - 1 from siti-tools produces some questionable values as dark scenes get very blown out, as seen in EXR dumps, I don't believe that's intended to be applied directly per pixel in that way.

Also add iPhone 14 Pro display model with both HDR and SDR flavors to test it. Since the content I have to test is vertical I also added vertical flavors of the display models (swap W and H) since I didn't see a better way to express the horizontal vs vertical rendering on mobile.

### CODE REVIEW

File: pycvvdp/display_model.py [hunk 1]
```diff
@@ -308,6 +331,11 @@ def forward( self, V ):
             L = pq2lin( V ).clip(0.005, self.Y_peak) + Y_black + Y_refl #TODO: soft clipping
         elif self.EOTF=='linear':
             L = V.clip(max(0.005, Y_black), self.Y_peak) + Y_refl #TODO: soft clipping
+        elif self.EOTF=='HLG':
+            gamma = 1.2
+            if self.Y_peak > 1000:
+                gamma = 1.2 + 0.42 * math.log10(self.Y_peak / 1000) - 0.07623 * math.log10(self.E_ambient / 5)
```
Comment (krasuluk) [inline pycvvdp/display_model.py:337]: @cosmin where does the correction term for ambient light come from? I do not see it in the rec 2100. Did I miss something?

File: pycvvdp/display_model.py [hunk 2]
```diff
@@ -308,6 +331,11 @@ def forward( self, V ):
             L = pq2lin( V ).clip(0.005, self.Y_peak) + Y_black + Y_refl #TODO: soft clipping
         elif self.EOTF=='linear':
             L = V.clip(max(0.005, Y_black), self.Y_peak) + Y_refl #TODO: soft clipping
+        elif self.EOTF=='HLG':
```
Comment (krasuluk) [inline pycvvdp/display_model.py:334]: I believe 'HLG' should also be added to the exception from clipping (at the line 322), as values below 0 and above 1 are allowed (see note 5g in the recommendation).

File: pycvvdp/display_model.py [hunk 3]
```diff
@@ -308,6 +331,11 @@ def forward( self, V ):
...
+            gamma = 1.2
+            if self.Y_peak > 1000:
+                gamma = 1.2 + 0.42 * math.log10(self.Y_peak / 1000) - 0.07623 * math.log10(self.E_ambient / 5)
+            L = (self.Y_peak-Y_black)*torch.pow(hlg2lin(V), gamma) + Y_black + Y_refl
```
Comment (krasuluk) [inline pycvvdp/display_model.py:338]: This implementation with this gamma is only correct when the EOTF is applied to the luma channel (Y). If we apply it to R, G, or B, it should be :

`L = (self.Y_peak-Y_black)*torch.pow(hlg2lin(Y), gamma - 1.0)*hlg2lin(V) + Y_black + Y_refl`

where Y = 0.2627 * R + 0.6780 * G + 0.0593 * B.

Note that if V = Y, the multiplication becomes `torch.pow(hlg2lin(Y), gamma - 1.0)*hlg2lin(Y)` which can be rewritten as `torch.pow(hlg2lin(Y), gamma - 1.0 + 1.0)`

File: pycvvdp/display_model.py (hunk 1 above)
Comment (cosmin) [reply pycvvdp/display_model.py:337]: I took this from note 5f

image

Comment (krasuluk) [reply pycvvdp/display_model.py:337]: @cosmin sorry, I wasn't entirely explicit -- I meant to ask about the correction term `- 0.07623 * math.log10(self.E_ambient / 5)`

Comment (cosmin) [reply pycvvdp/display_model.py:337]: @krasuluk ah, so BT.2100 says

> Note 5g – The system gamma value may be decreased for brighter background and surround conditions.

But doesn't specify how to decrease the gamma for bright ambient conditions. However it turns out the BBC ran some experiments on this described in BBC Research & Development White Paper WHP 369 https://downloads.bbc.co.uk/rd/pubs/whp/whp-pdf-files/WHP369.pdf

Here's the relevant part

image

This is also captured in ITU-R BT.2390-10 which quotes the BBC experiment

> Recommendation ITU-R BT.2100 recognises that the HLG display gamma may need to be reduced in brighter viewing environments, to compensate for the differences in the adaptation state of the eye. The BBC conducted subjective tests to measure the change in gamma necessary to perceptually match images displayed across a range of peak luminances in the reference and in non-reference environments. Twenty-one viewers participated in the tests.

image

I kept the 0.07623 from the original BBC white paper, but we could round that to 0.076. 

BT.2390-10 also mentions another option 

image

from "Display of high dynamic range images under varying viewing conditions" http://dx.doi.org/10.1117/12.2274253 but I didn't investigate that further.

Comment (krasuluk) [reply pycvvdp/display_model.py:337]: Thanks so much for the detailed explanation and links @cosmin 🙏 all is clear now 👍

File: pycvvdp/display_model.py (hunk 3 above)
Comment (cosmin) [reply pycvvdp/display_model.py]: after doing some digging I confirmed that by the time we enter the forward functions we are in the RGB domain, therefore I implemented the RGB formulation of the EOTF, using the `Y**(gamma-1) * rgb`.

Comment (krasuluk) [reply pycvvdp/display_model.py]: LGTM! One nit -- would it be worth adding a check for the input being 3-dimensional (RGB) to make sure nobody tries to use it per channel? I am not sure if the code is permissive of such usage but I always prefer to reinforce the implementation. Just a soft suggestion.

**Comment by cosmin** [comment]
@mantiuk any feedback on this pull request?

**Review by mantiuk — APPROVED** [review]
Everything looks very good. 

Sorry for taking so long to merge with the main branch. I was preoccupied with a few other things.

**Pull request merged**